perf(transformer/styled-components): simplify CSS minification

overlookmotel opened the pull request:
Simplify CSS minification.

* Remove the step of adding placeholders, which means we also don't need to remove them again, and don't need a hashmap to track which quasis are retained. Instead, just work through the quasis one by one.
* Remove the intermediate `new_raws` `Vec`.
* Skip `\\`, `\'`, `\"` in byte search loop, to avoid backtracking to check for `\"` when exiting a string.

The whole thing squashes down to a single function.

I *believe* I've kept the behavior exactly as it was before. Is there anything I'm missing?

There's one bit of the logic I don't understand. `/* ... */` is replaced with a space, unless there's a space preceding / following it. But I don't think we need to add a space in cases like `height:/* big */${10000}px`. Do we?

overlookmotel commented on crates/oxc_transformer/src/plugins/styled_components.rs:
```diff
@@ -861,258 +824,225 @@ fn is_valid_styled_component_source(source: &str) -> bool {
+                                        comment_type = Some(true);
+                                        break;
+                                    }
+                                    Some(&b) if b != b'!' => {
+                                        let end_index =
+                                            bytes[i + 2..].windows(2).position(|q| q == b"*/");
+                                        if let Some(end_index) = end_index {
+                                            i += end_index + 4;
+
+                                            // Add a space when this is a own line block comment
+                                            if i < bytes.len()
+                                                && !bytes[i].is_ascii_whitespace()
+                                                && output.last().is_some_and(|&last| last != b' ')
+                                            {
+                                                output.push(b' ');
+                                            }
+                                            continue;
+                                        }
+
+                                        // Comment is not complete at end of quasi
+                                        comment_type = Some(true);
+                                        break;
+                                    }
+                                    _ => {}
+                                }
```
It doesn't matter.

overlookmotel commented on crates/oxc_transformer/src/plugins/styled_components.rs:
```diff
@@ -861,258 +824,229 @@ fn is_valid_styled_component_source(source: &str) -> bool {
+                            // Skip line comments, but be careful not to break URLs like `https://...`
+                            b'/' if paren_depth == 0 && (i == 0 || bytes[i - 1] != b':') => {
+                                let end_index =
+                                    bytes[i + 2..].iter().position(|&b| matches!(b, b'\n' | b'\r'));
+                                if let Some(end_index) = end_index {
+                                    i += end_index + 2; // On `\n` or `\r`
+                                    continue;
```
`i += end_index + 2` and `i = i + 2 + end_index` are identical!

Dunqing commented:
> There's one bit of the logic I don't understand. `/* ... */` is replaced with a space, unless there's a space preceding / following it. But I don't think we need to add a space in cases like `height:/* big */${10000}px`. Do we?

This is what the `Babel` plugin does, I don't know why.

https://github.com/styled-components/babel-plugin-styled-components/blob/4e2eb388d9c90f2921c306c760657d059d01a518/src/minify/index.js#L91-L92

overlookmotel committed:
perf(transformer/styled-components): simplify CSS minification (#12224)

Simplify CSS minification.

* Remove the step of adding placeholders, which means we also don't need to remove them again, and don't need a hashmap to track which quasis are retained. Instead, just work through the quasis one by one.
* Remove the intermediate `new_raws` `Vec`.
* Skip `\\`, `\'`, `\"` in byte search loop, to avoid backtracking to check for `\"` when exiting a string.

The whole thing squashes down to a single function.

I *believe* I've kept the behavior exactly as it was before. Is there anything I'm missing?

There's one bit of the logic I don't understand. `/* ... */` is replaced with a space, unless there's a space preceding / following it. But I don't think we need to add a space in cases like `height:/* big */${10000}px`. Do we?
